# Análisis de base de datos `olist`

## Importación de paquetes

In [10]:
import sqlite3

import pandas as pd
from geopy.distance import geodesic
from IPython.display import display

## Obtención del _dataset_

Con el código a continuación obtenemos el _dataset_ a partir de la base de datos _SQLite_ `olist.db`. El _dataset_ contiene información de órdenes, clientes, vendedores y productos, junto con datos geográficos asociados a códigos postales.

Para más información sobre esta base datos ver <https://www.kaggle.com/datasets/olistbr/brazilian-ecommerce>.

El archivo `olist.db` es una conversión de los archivos CSV originales a una base de datos SQLite. Esta conversión se realizó para facilitar la consulta y manipulación de los datos.

In [11]:
with sqlite3.connect("data/olist.db") as conn:
    query = """
    -- Obtenemos, según cada código postal, la latitud y longitud promedio.
    WITH geo_agg AS (
        SELECT 
            geolocation_zip_code_prefix,
            AVG(geolocation_lat) as lat,
            AVG(geolocation_lng) as lng
        FROM geolocation
        GROUP BY geolocation_zip_code_prefix
    ),
    -- Obtenemos, para cada orden, una agregación de todos los items incluidos en ella.
    item_agg AS (
        SELECT
            i.order_id,
            SUM(i.price) as total_price,
            SUM(i.freight_value) as total_freight_value,
            SUM(p.product_weight_g) as total_weight_g,
            SUM(p.product_length_cm * p.product_height_cm * p.product_width_cm) as total_volume_cm3,
            COUNT(*) as item_count,
            -- Hay ordenes con múltiples sellers, para ellos calculamos la latitud y longitud promedio.
            -- No es ideal pero es un compromiso razonable.
            AVG(geo_seller.lat) as avg_lat_seller,
            AVG(geo_seller.lng) as avg_lng_seller,
            -- Para el estado del seller también hacemos un compromiso agarrando el primero en orden alfabético.
            MIN(s.seller_state) as seller_state
        FROM order_items i
        JOIN products p ON i.product_id = p.product_id
        JOIN sellers s ON i.seller_id = s.seller_id
        LEFT JOIN geo_agg geo_seller ON s.seller_zip_code_prefix = geo_seller.geolocation_zip_code_prefix
        -- Filtramos los productos que no tengan datos de peso o dimensiones.
        WHERE p.product_weight_g IS NOT NULL
          AND p.product_length_cm IS NOT NULL
          AND p.product_height_cm IS NOT NULL
          AND p.product_width_cm IS NOT NULL
        GROUP BY i.order_id
    )
    SELECT
        o.order_id,
        o.order_purchase_timestamp,
        o.order_delivered_customer_date,
        a.total_price,
        a.total_freight_value,
        a.total_weight_g,
        a.total_volume_cm3,
        a.item_count,
        c.customer_state,
        a.seller_state,
        geo_customer.lat as lat_customer,
        geo_customer.lng as lng_customer,
        a.avg_lat_seller as lat_seller,
        a.avg_lng_seller as lng_seller
    FROM orders o
    JOIN item_agg a ON o.order_id = a.order_id
    JOIN customers c ON o.customer_id = c.customer_id
    LEFT JOIN geo_agg geo_customer ON c.customer_zip_code_prefix = geo_customer.geolocation_zip_code_prefix
    -- Solo nos interesan las órdenes que fueron entregadas.
    WHERE o.order_status = 'delivered'
      -- También filtramos las ordenes que no tengan ciertos datos.
      AND o.order_delivered_customer_date IS NOT NULL
      AND geo_customer.lat IS NOT NULL
      AND a.avg_lat_seller IS NOT NULL
    """

    df = pd.read_sql(query, conn)

df["order_purchase_timestamp"] = pd.to_datetime(df["order_purchase_timestamp"])
df["order_delivered_customer_date"] = pd.to_datetime(df["order_delivered_customer_date"])

# Calculamos la distancia entre el cliente y el seller mediante distancia geodésica.
df["distance_km"] = df.apply(
    lambda row: geodesic((row["lat_customer"], row["lng_customer"]), (row["lat_seller"], row["lng_seller"])).kilometers,
    axis=1,
)

display(df.head())
display(df.info())

,order_id,order_purchase_timestamp,order_delivered_customer_date,total_price,total_freight_value,total_weight_g,total_volume_cm3,item_count,customer_state,seller_state,lat_customer,lng_customer,lat_seller,lng_seller,distance_km
0,00010242fe8c5a6d1ba2dd792cb16214,2017-09-13 08:59:02,2017-09-20 23:43:48,58.90,13.29,650.0,3528.0,1,RJ,SP,-21.762775,-41.309633,-22.496953,-44.127492,301.858959
1,00018f77f2f0320c557190d7a144bdd3,2017-04-26 10:53:06,2017-05-12 16:04:24,239.90,19.93,30000.0,60000.0,1,SP,SP,-20.220527,-50.903424,-23.565096,-46.518565,585.131104
2,000229ec398224ef6ca0657da4fc703e,2018-01-14 14:33:31,2018-01-22 13:19:16,199.00,17.87,3050.0,14157.0,1,MG,MG,-19.870305,-44.593326,-22.262584,-46.171124,311.506212
3,00024acbcdf0a6daa1e931b038114c75,2018-08-08 10:00:35,2018-08-14 13:32:39,12.99,12.79,200.0,2400.0,1,SP,SP,-23.089925,-46.611654,-20.553624,-47.387359,292.064162
4,00042b26cf59d7ce69dfabb4e55b4fd9,2017-02-04 13:57:51,2017-03-01 16:42:31,199.90,18.14,3750.0,42000.0,1,SP,PR,-23.243402,-46.827614,-22.929384,-53.135873,647.209811


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 95978 entries, 0 to 95977
Data columns (total 15 columns):
 #   Column                         Non-Null Count  Dtype         
---  ------                         --------------  -----         
 0   order_id                       95978 non-null  object        
 1   order_purchase_timestamp       95978 non-null  datetime64[ns]
 2   order_delivered_customer_date  95978 non-null  datetime64[ns]
 3   total_price                    95978 non-null  float64       
 4   total_freight_value            95978 non-null  float64       
 5   total_weight_g                 95978 non-null  float64       
 6   total_volume_cm3               95978 non-null  float64       
 7   item_count                     95978 non-null  int64         
 8   customer_state                 95978 non-null  object        
 9   seller_state                   95978 non-null  object        
 10  lat_customer                   95978 non-null  float64       
 11  lng_customer   

None

Con este último `df.info()` podemos ver que se obtuvo el _dataset_ de forma correcta, con un total de 95978 filas y 15 columnas.